### DISCUSSION TITLE
Sub image search

### DISCUSSION THREAD

**Discussion by uuf6429** [libvips/libvips#3528] [Q&A] [answered]
I'm trying to use libvips (initially from cli) to solve this problem: https://stackoverflow.com/questions/76416455/searching-for-image-within-another-one

Unfortunately, I'm very new to vips and image processing, so I'm quite clueless at this point. I did look at [the api index](https://www.libvips.org/API/current/api-index-full.html) as well as `vips -h` and `vips -l`.

PS: To be clear, I'm not looking for some ML or fancy AI-based image processing.
PPS: I already saw this, but I don't think it relates to my problem: https://github.com/libvips/libvips/issues/1923

**Comment by jcupitt** [comment, marked as answer]
Hi @uuf6429,

I'd use `fastcor`. You give it a big image and a small one, it positions the small image at every possible position in the big one, and computes sum of squares of differences. You get an image back where each pixel has the value of SoSoD for the subimage at that point. You'd then need to search this result image and find the minimum (ie. the smallest difference is the best match).

For example, if I have this test image:

![nina](https://github.com/libvips/libvips/assets/580843/0b628b0a-97e6-46da-9ff5-6375a9b2694b)

I can cut out part of it like this:

```
$ vips crop nina.jpg eye.jpg 2260 1840 256 256
```

To make:

![eye](https://github.com/libvips/libvips/assets/580843/f48872bd-cb08-4798-9fdf-3e3f186c84d2)

Searching a large colour image for a 256x256 colour image will be very slow, so it's best to make low-res black and white images for a quick rough search.

```
$ vips colourspace nina.jpg nina-mono.jpg b-w
$ vips colourspace eye.jpg eye-mono.jpg b-w
$ vips resize nina-mono.jpg small-nina-mono.jpg 0.0625
$ vips resize eye-mono.jpg small-eye-mono.jpg 0.0625
```

Then search for it like this:

```
$ vips fastcor small-nina-mono.jpg small-eye-mono.jpg x.v
```

You need the position of the minimum:

```
$ vips min x.v --x --y
149
123
40294.000000
```

We shrank by x16, so subtract 8 (half the size of the subimage) and scale up those coordinates to get (2256, 1840), very close to the original crop (x, y).

It's simpler in php of course, perhaps:

```php
$ref = $big->colourspace("b-w")->resize(0.0625);
$sec = $small->colourspace("b-w")->resize(0.0625);
$match = $big->fastcor($small);
$x, $y, $min = $match->min(["x" -> true, "y" => true]);
```

**Reply by uuf6429** [reply]
Hey, many thanks for the very detailed reply!

**Comment by uuf6429** [comment]
Hey, would you recommend any particular troubleshooting steps in case something isn't working as expected?

I have this particular case ([Archive.zip](https://github.com/libvips/libvips/files/11721672/Archive.zip)) that isn't working, but I'd like to know what's the best way to troubleshoot it (so far I tried changing/disabling the resize and colorspace values).

**Reply by jcupitt** [reply]
I usually develop this stuff in nip2, then code it up in python once it seems to be working.

https://github.com/libvips/nip2

It lets you watch all the pipelines evaluate, you can check pixels as they are processed and tweak any parameters. You can run nip2 workspace as scripts too, though a python version is usually more useful.

**Reply by jcupitt** [reply]
eg.:

![image](https://github.com/libvips/libvips/assets/580843/c7174f9f-5e32-498e-be97-e022ca3a4d93)

http://www.rollthepotato.net/~john/christian.ws

You can drag and drop test images into A1 and it'll all recompute, or right click on A1, select "replace from file", pin up the file dialog, and try running on a directory of test images.

**Reply by uuf6429** [reply]
I've managed to get it to compile on osx.
Something odd: I had to manually `cp -R ./share/nip2 /usr/local/Cellar/vips/8.14.2_1/share` to make the app work correctly.

Anyway, I was not able to load the png:
![image](https://github.com/libvips/libvips/assets/230049/bb7c5fb2-03b0-4b62-bfe2-04fc2461adf3)
```bash
~ vips --vips-config
enable debug: false
enable deprecated: true
enable modules: true
enable cplusplus: true
enable RAD load/save: true
enable Analyze7 load/save: true
enable PPM load/save: true
enable GIF load: true
use fftw for FFTs: true
accelerate loops with ORC: true
ICC profile support with lcms: true
zlib: true
text rendering with pangocairo: true
font file support with fontconfig: true
EXIF metadata support with libexif: true
JPEG load/save with libjpeg: true
JXL load/save with libjxl: true (dynamic module: true)
JPEG2000 load/save with OpenJPEG: true
PNG load/save with libspng: true
PNG load/save with libpng: false
selected quantisation package: imagequant
TIFF load/save with libtiff: true
image pyramid save with libgsf: true
HEIC/AVIF load/save with libheif: true (dynamic module: true)
WebP load/save with libwebp: true
PDF load with PDFium: false
PDF load with poppler-glib: true (dynamic module: true)
SVG load with librsvg: true
EXR load with OpenEXR: true
OpenSlide load: true (dynamic module: true)
Matlab load with libmatio: true
NIfTI load/save with niftiio: false
FITS load/save with cfitsio: true
GIF save with cgif: true
selected Magick package: MagickCore (dynamic module: true)
Magick API version: magick7
Magick load: true
Magick save: true

~  head -c 10 ./Downloads/screenshot.png
�PNG
```

**Reply by jcupitt** [reply]
Oh, well done.

You need to use libpng, not libspng, for nip2. I ought to fix this.

edit: ooop, I forgot, I fixed this a few months ago.  libvips 8.15 will work with nip2 even with ilbspng.

**Reply by uuf6429** [reply]
So what shall I try/do next? I built the app from the master branch and I also have both libpng and libspng installed.

**Reply by jcupitt** [reply]
You can disable libspng when you build libvips with eg.:

```
$ meson setup build --prefix ~/vips -Dspng=disabled
```

That'll make it fall back to libpng. You could also build libvips from master, which has a fix for this issue.

**Reply by uuf6429** [reply]
> build libvips from master, which has a fix for this issue.

aah, understood - I didn't get that from your previous reply. 👍
